src/mscorlib/src/System/Lazy.cs

```diff
@@ -27,12 +27,160 @@
+        internal readonly static LazyHelper PublicationOnlyViaConstructor = new LazyHelper(LazyState.PublicationOnlyViaConstructor);
+        internal readonly static LazyHelper PublicationOnlyViaFactory     = new LazyHelper(LazyState.PublicationOnlyViaFactory);
+        internal readonly static LazyHelper PublicationOnlyWaitForOtherThreadToPublish       = new LazyHelper(LazyState.PublicationOnlyWait);
+
+        internal LazyState State { get; }
+
+        private readonly ExceptionDispatchInfo _exceptionDispatch;
+
+        /// <summary>
+        /// Constructor that defines the state
+        /// </summary>
+        internal LazyHelper(LazyState state)
+        {
+            State = state;
+        }
+
+        /// <summary>
+        /// Constructor used for exceptions
+        /// </summary>
+        internal LazyHelper(LazyThreadSafetyMode mode, Exception exception)
+        {
+            switch(mode)
+            {
+                case LazyThreadSafetyMode.ExecutionAndPublication:
+                    State = LazyState.ExecutionAndPublicationException;
+                    break;
+
+                case LazyThreadSafetyMode.None:
+                    State = LazyState.NoneException;
+                    break;
+
+                case LazyThreadSafetyMode.PublicationOnly:
+                    State = LazyState.PublicationOnlyException;
+                    break;
+
+                default:
+                    Debug.Assert(false, "internal constructor, this should never occur");
+                    break;
+            }
+
+            _exceptionDispatch = ExceptionDispatchInfo.Capture(exception);
+        }
+
+        internal void ThrowException()
+        {
+            if (_exceptionDispatch == null)
+            {
+                Debug.Assert(false, "execution path is invalid");
+            }
+
+            _exceptionDispatch.Throw();
+        }
+
+        private LazyThreadSafetyMode GetMode()
+        {
+            switch (State)
+            {
+                case LazyState.NoneViaConstructor:
+                case LazyState.NoneViaFactory:
+                case LazyState.NoneException:
+                    return LazyThreadSafetyMode.None;
+
+                case LazyState.PublicationOnlyViaConstructor:
+                case LazyState.PublicationOnlyViaFactory:
+                case LazyState.PublicationOnlyWait:
+                case LazyState.PublicationOnlyException:
+                    return LazyThreadSafetyMode.PublicationOnly;
+
+                case LazyState.ExecutionAndPublicationViaConstructor:
+                case LazyState.ExecutionAndPublicationViaFactory:
+                case LazyState.ExecutionAndPublicationException:
+                    return LazyThreadSafetyMode.ExecutionAndPublication;
+
+                default:
+                    Debug.Assert(false, "Invalid logic; State should always have a valid value");
+                    throw new ArgumentOutOfRangeException();
```

Comment:
You can just return `return default(LazyThreadSafetyMode)` to make the compiler happy - it is smaller/faster code.